File: todolist/src/main/java/org/com/todolist/global/exception/GlobalExceptionHandler.java
package org.com.todolist.global.exception;

import java.util.HashMap;
import java.util.Map;
import org.com.todolist.global.api.ApiResponse;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * 검증/도메인 예외를 일괄 처리.
 */
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(TodoNotFoundException.class)
    public ResponseEntity<ApiResponse<Object>> handleNotFound(TodoNotFoundException ex) {
        return ResponseEntity.status(HttpStatus.NOT_FOUND)
            .body(ApiResponse.error(ex.getMessage()));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiResponse<Object>> handleValidation(MethodArgumentNotValidException ex) {
        Map<String, String> errors = new HashMap<>();
        for (FieldError fe : ex.getBindingResult().getFieldErrors()) {
            errors.put(fe.getField(), fe.getDefaultMessage());
        }
        return ResponseEntity.badRequest()
            .body(ApiResponse.error("Validation failed: " + errors));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiResponse<Object>> handleOthers(Exception ex) {
        // 운영 환경에선 상세 메시지 노출 제한
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(ApiResponse.error(ex.getMessage()));
    }

}

File: ssr-board/src/main/kotlin/org/com/ssrboard/global/error/GlobalExceptionHandler.kt
package org.com.ssrboard.global.error

import jakarta.servlet.http.HttpServletResponse
import jakarta.validation.ConstraintViolationException
import org.com.ssrboard.global.web.ApiResponse
import org.springframework.http.HttpStatus
import org.springframework.ui.Model
import org.springframework.validation.BindException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody

@ControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(EntityNotFoundException::class)
    fun notFound(ex: EntityNotFoundException, resp: HttpServletResponse): String {
        resp.status = HttpStatus.NOT_FOUND.value()
        return "error/404"
    }

    @ResponseBody
    @ExceptionHandler(BusinessException::class)
    fun business(ex: BusinessException): ApiResponse<Nothing> =
        ApiResponse.fail(ex.message ?: "Business error")

    @ExceptionHandler(Exception::class)
    fun any(ex: Exception, resp: HttpServletResponse): String {
        resp.status = HttpStatus.INTERNAL_SERVER_ERROR.value()
        return "error/500"
    }

}